logpy/tests/test_sudoku.py

```diff
@@ -0,0 +1,164 @@
+        else:
+            return var()
+    return tuple(helper(x) for x in hints)
+
+
+def all_numbers(coll):
+    return permuteq(coll, DIGITS)
+
+
+def sudoku_solver(hints):
+    variables = vars(hints)
+    rows = get_rows(variables)
+    cols = get_columns(rows)
+    sqs = get_squares(rows)
+    return run(
+        1,
+        variables,
+        everyg(all_numbers, rows),
+        everyg(all_numbers, cols),
+        everyg(all_numbers, sqs)
+    )
+
+
+def test_missing_one_entry():
+    example_board = (
+        5, 3, 4, 6, 7, 8, 9, 1, 2,
+        6, 7, 2, 1, 9, 5, 3, 4, 8,
+        1, 9, 8, 3, 4, 2, 5, 6, 7,
+        8, 5, 9, 7, 6, 1, 4, 2, 3,
+        4, 2, 6, 8, 5, 3, 7, 9, 1,
+        7, 1, 3, 9, 2, 4, 8, 5, 6,
+        9, 6, 1, 5, 3, 7, 2, 8, 4,
+        2, 8, 7, 4, 1, 9, 6, 3, 5,
+        3, 4, 5, 2, 8, 6, 0, 7, 9
+    )
+    expected_solution = (
+        5, 3, 4, 6, 7, 8, 9, 1, 2,
+        6, 7, 2, 1, 9, 5, 3, 4, 8,
+        1, 9, 8, 3, 4, 2, 5, 6, 7,
+        8, 5, 9, 7, 6, 1, 4, 2, 3,
+        4, 2, 6, 8, 5, 3, 7, 9, 1,
+        7, 1, 3, 9, 2, 4, 8, 5, 6,
+        9, 6, 1, 5, 3, 7, 2, 8, 4,
+        2, 8, 7, 4, 1, 9, 6, 3, 5,
+        3, 4, 5, 2, 8, 6, 1, 7, 9,
+    )
+    assert sudoku_solver(example_board)[0] == expected_solution
+
+
+def test_missing_complex_board():
+    example_board = (
+        5, 3, 4, 6, 7, 8, 9, 0, 2,
+        6, 7, 2, 0, 9, 5, 3, 4, 8,
+        0, 9, 8, 3, 4, 2, 5, 6, 7,
+        8, 5, 9, 7, 6, 0, 4, 2, 3,
+        4, 2, 6, 8, 5, 3, 7, 9, 0,
+        7, 0, 3, 9, 2, 4, 8, 5, 6,
+        9, 6, 0, 5, 3, 7, 2, 8, 4,
+        2, 8, 7, 4, 0, 9, 6, 3, 5,
+        3, 4, 5, 2, 8, 6, 0, 7, 9,
+    )
+    expected_solution = (
+        5, 3, 4, 6, 7, 8, 9, 1, 2,
+        6, 7, 2, 1, 9, 5, 3, 4, 8,
+        1, 9, 8, 3, 4, 2, 5, 6, 7,
+        8, 5, 9, 7, 6, 1, 4, 2, 3,
+        4, 2, 6, 8, 5, 3, 7, 9, 1,
+        7, 1, 3, 9, 2, 4, 8, 5, 6,
+        9, 6, 1, 5, 3, 7, 2, 8, 4,
+        2, 8, 7, 4, 1, 9, 6, 3, 5,
+        3, 4, 5, 2, 8, 6, 1, 7, 9,
+    )
+    assert sudoku_solver(example_board)[0] == expected_solution
+
+
+def test_insolvable():
+    example_board = (
+        5, 3, 4, 6, 7, 8, 9, 1, 2,
+        6, 7, 2, 1, 9, 5, 9, 4, 8,  # Note column 7 has two 9's.
+        1, 9, 8, 3, 4, 2, 5, 6, 7,
+        8, 5, 9, 7, 6, 1, 4, 2, 3,
+        4, 2, 6, 8, 5, 3, 7, 9, 1,
+        7, 1, 3, 9, 2, 4, 8, 5, 6,
+        9, 6, 1, 5, 3, 7, 2, 8, 4,
+        2, 8, 7, 4, 1, 9, 6, 3, 5,
+        3, 4, 5, 2, 8, 6, 0, 7, 9
+    )
+    assert sudoku_solver(example_board) == ()
+
+
+def test_many_missing_elements():
+    example_board = (
+        5, 3, 0, 0, 7, 0, 0, 0, 0,
+        6, 0, 0, 1, 9, 5, 0, 0, 0,
+        0, 9, 8, 0, 0, 0, 0, 6, 0,
+        8, 0, 0, 0, 6, 0, 0, 0, 3,
+        4, 0, 0, 8, 0, 3, 0, 0, 1,
+        7, 0, 0, 0, 2, 0, 0, 0, 6,
+        0, 6, 0, 0, 0, 0, 2, 8, 0,
+        0, 0, 0, 4, 1, 9, 0, 0, 5,
+        0, 0, 0, 0, 8, 0, 0, 7, 9)
+    assert sudoku_solver(example_board)[0] == (
```

Comment:
The good news is that the bug found in https://github.com/holtchesley/embedded-logic/blob/master/logpy/sudoku.ipynb seems to have been fixed, though solving it is now very slow.